File: backend/utils/mpesaService.js
const axios = require('axios');
const logger = require('./logger');

class MpesaService {
  constructor() {
    this.consumerKey = process.env.MPESA_CONSUMER_KEY;
    this.consumerSecret = process.env.MPESA_CONSUMER_SECRET;
    this.shortCode = process.env.MPESA_SHORTCODE || process.env.MPESA_PAYBILL_OR_SHORTCODE;
    this.passkey = process.env.MPESA_PASSKEY;
    this.callbackUrl = process.env.MPESA_CALLBACK_URL;
    this.baseUrl = (process.env.MPESA_ENVIRONMENT || process.env.MPESA_ENV) === 'production'
      ? 'https://api.safaricom.co.ke'
      : 'https://sandbox.safaricom.co.ke';
    this.isMock = process.env.MPESA_MOCK === 'true' || process.env.NODE_ENV === 'test';

    // Production safety guard — fail fast if mock is on in prod
    if (this.isMock && process.env.NODE_ENV === 'production') {
      throw new Error('[FATAL] MPESA_MOCK=true is not allowed in production. Unset it before deploying.');
    }
    if (this.isMock) {
      logger.warn('⚠️  M-Pesa MOCK MODE is ACTIVE — no real M-Pesa transactions will occur');
    }
  }

  /**
   * Get OAuth Access Token from Daraja
   */
  async getAccessToken() {
    try {
      if (this.isMock) {
        logger.info('M-Pesa Mock: Generating access token');
        return 'mock_access_token_' + Date.now();
      }

      const auth = Buffer.from(`${this.consumerKey}:${this.consumerSecret}`).toString('base64');
      const response = await axios.get(`${this.baseUrl}/oauth/v1/generate?grant_type=client_credentials`, {
        headers: {
          Authorization: `Basic ${auth}`,
        },
        timeout: 10000,
      });
      return response.data.access_token;
    } catch (error) {
      logger.error('M-Pesa OAuth failed', { error: error.response?.data || error.message });
      // Preserve the error code and response data for the controller
      const authError = new Error(error.message || 'Failed to authenticate with M-Pesa');
      authError.code = error.code;
      authError.response = error.response;
      throw authError;
    }
  }

  /**
   * Initiate STK Push (Lipa Na M-Pesa Online)
   */
  async initiateStkPush(phoneNumber, amount, accountReference, transactionDesc) {
    try {
      const token = await this.getAccessToken();
      const timestamp = new Date().toISOString().replace(/[-:T.Z]/g, '').slice(0, 14);
      const password = Buffer.from(`${this.shortCode}${this.passkey}${timestamp}`).toString('base64');

      // Normalize phone number to 2547XXXXXXXX
      let formattedPhone = phoneNumber.replace(/\+/g, '');
      if (formattedPhone.startsWith('0')) formattedPhone = '254' + formattedPhone.slice(1);
      if (formattedPhone.startsWith('7') || formattedPhone.startsWith('1')) formattedPhone = '254' + formattedPhone;

      const payload = {
        BusinessShortCode: this.shortCode,
        Password: password,
        Timestamp: timestamp,
        TransactionType: 'CustomerPayBillOnline',
        Amount: Math.round(amount),
        PartyA: formattedPhone,
        PartyB: this.shortCode,
        PhoneNumber: formattedPhone,
        CallBackURL: this.callbackUrl,
        AccountReference: accountReference.substring(0, 12),
        TransactionDesc: transactionDesc.substring(0, 20),
      };

      if (this.isMock) {
        logger.info('M-Pesa Mock: Initiating STK push', { phoneNumber, amount });
        // Simulate a small delay
        await new Promise(resolve => setTimeout(resolve, 800));
        return {
          MerchantRequestID: 'mock_merchant_' + Date.now(),
          CheckoutRequestID: 'mock_checkout_' + Date.now(),
          ResponseCode: '0',
          ResponseDescription: 'Success. Request accepted for processing',
          CustomerMessage: 'Success. Request accepted for processing',
          isMock: true
        };
      }

      const response = await axios.post(`${this.baseUrl}/mpesa/stkpush/v1/processrequest`, payload, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
        timeout: 10000,
      });

      return response.data;
    } catch (error) {
      logger.error('STK Push initiation failed', { error: error.response?.data || error.message });
      throw error;
    }
  }

  /**
   * Query STK Push status (Fallback handle)
   */
  async queryStatus(checkoutRequestId) {
    try {
      const token = await this.getAccessToken();
      const timestamp = new Date().toISOString().replace(/[-:T.Z]/g, '').slice(0, 14);
      const password = Buffer.from(`${this.shortCode}${this.passkey}${timestamp}`).toString('base64');

      const payload = {
        BusinessShortCode: this.shortCode,
        Password: password,
        Timestamp: timestamp,
        CheckoutRequestID: checkoutRequestId,
      };

      if (this.isMock) {
        logger.info('M-Pesa Mock: Querying status', { checkoutRequestId });
        return {
          ResponseCode: '0',
          ResponseDescription: 'The service request has been accepted successfully',
          MerchantRequestID: 'mock_merchant_' + Date.now(),
          CheckoutRequestID: checkoutRequestId,
          ResultCode: '0',
          ResultDesc: 'The service request is processed successfully.',
          isMock: true
        };
      }

      const response = await axios.post(`${this.baseUrl}/mpesa/stkpushquery/v1/query`, payload, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
        timeout: 10000,
      });

      return response.data;
    } catch (error) {
       logger.error('STK status query failed', { checkoutRequestId, error: error.response?.data || error.message });
       throw error;
    }
  }

  /**
   * Simulate an M-Pesa callback payload for mock mode testing.
   * Enables full E2E dev testing without a live Safaricom connection.
   */
  async simulateMockCallback(checkoutRequestId, amount, phoneNumber) {
    if (!this.isMock) throw new Error('simulateMockCallback is only available in mock mode');
    logger.info('M-Pesa Mock: Simulating Safaricom callback', { checkoutRequestId });
    await new Promise(r => setTimeout(r, 500)); // Simulate async timing
    return {
      Body: {
        stkCallback: {
          MerchantRequestID: 'mock_merchant_' + Date.now(),
          CheckoutRequestID: checkoutRequestId,
          ResultCode: 0,
          ResultDesc: 'The service request is processed successfully.',
          CallbackMetadata: {
            Item: [
              { Name: 'Amount', Value: parseFloat(amount) },
              { Name: 'MpesaReceiptNumber', Value: 'MOCK' + Date.now() },
              { Name: 'TransactionDate', Value: Date.now() },
              { Name: 'PhoneNumber', Value: phoneNumber }
            ]
          }
        }
      }
    };
  }
}

module.exports = new MpesaService();

File: backend/src/main/java/com/chamasmart/backend/service/MpesaService.java
package com.chamasmart.backend.service;
import lombok.extern.slf4j.Slf4j;
import com.chamasmart.backend.domain.*;
import com.chamasmart.backend.dto.*;
import com.chamasmart.backend.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.math.BigDecimal;
import java.time.ZonedDateTime;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.ApplicationContext;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

@Service
@RequiredArgsConstructor
@Slf4j
public class MpesaService {
    private final MpesaTransactionRepository mpesaTransactionRepository;
    private final ContributionRepository contributionRepository;
    private final ChamaRepository chamaRepository;
    private final UserRepository userRepository;
    private final ChamaMemberRepository chamaMemberRepository;
    private final FinancialAuditLogRepository auditLogRepository;
    private final ApplicationContext applicationContext;

    @Value("${app.mpesa.env:sandbox}")
    private String env;

    @Value("${app.mpesa.consumer-key:cpQnnxLsrtEU7WQ94G5XwfBGG11yUckjQAnAJPQXM9JkHycV}")
    private String consumerKey;

    @Value("${app.mpesa.consumer-secret:YOUR_DARAJA_SANDBOX_CONSUMER_SECRET}")
    private String consumerSecret;

    @Value("${app.mpesa.shortcode:174379}")
    private String shortcode;

    @Value("${app.mpesa.passkey:bfb279f9aa9bdbcf158e97dd71a467cd2e0c893059b10f78e6b72ada1ed2c919}")
    private String passkey;

    @Value("${app.mpesa.callback-url:https://chamasmart-xcym.onrender.com/api/v1/mpesa/callback}")
    private String callbackUrl;

    private final RestTemplate restTemplate = new RestTemplate();

    private String getAccessToken() {
        String baseUrl = "production".equalsIgnoreCase(env) ? "https://api.safaricom.co.ke" : "https://sandbox.safaricom.co.ke";
        String url = baseUrl + "/oauth/v1/generate?grant_type=client_credentials";

        String auth = consumerKey + ":" + consumerSecret;
        byte[] encodedAuth = java.util.Base64.getEncoder().encode(auth.getBytes(java.nio.charset.StandardCharsets.UTF_8));
        String authHeader = "Basic " + new String(encodedAuth);

        HttpHeaders headers = new HttpHeaders();
        headers.set("Authorization", authHeader);
        HttpEntity<String> entity = new HttpEntity<>(headers);

        try {
            ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.GET, entity, Map.class);
            if (response.getStatusCode().is2xxSuccessful() && response.getBody() != null) {
                return (String) response.getBody().get("access_token");
            }
        } catch (Exception e) {
            log.error("Failed to get Daraja access token: {}", e.getMessage());
        }
        return null;
    }

    private String getPassword(String timestamp) {
        String str = shortcode + passkey + timestamp;
        return java.util.Base64.getEncoder().encodeToString(str.getBytes(java.nio.charset.StandardCharsets.UTF_8));
    }

    @Transactional
    public MpesaTransaction initiateStkPush(MpesaStkPushRequestDto requestDto, Long userId) {
        log.info("Initiating M-Pesa STK Push for user ID: {}, chama ID: {}, amount: {}", userId, requestDto.getChamaId(), requestDto.getAmount());
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        Chama chama = chamaRepository.findById(requestDto.getChamaId())
                .orElseThrow(() -> new RuntimeException("Chama not found"));
        // Check if we should use mock logic
        if (consumerSecret.contains("YOUR_DARAJA")) {
            log.info("Using MOCKED Daraja STK Push because consumer-secret is a placeholder.");
            String checkoutRequestId = "ws_CO_" + UUID.randomUUID().toString().substring(0, 15).replace("-", "");
            String merchantRequestId = "req_" + UUID.randomUUID().toString().substring(0, 10).replace("-", "");
            
            MpesaTransaction transaction = MpesaTransaction.builder()
                    .checkoutRequestId(checkoutRequestId)
                    .merchantRequestId(merchantRequestId)
                    .user(user)
                    .chama(chama)
                    .amount(requestDto.getAmount())
                    .phoneNumber(requestDto.getPhoneNumber())
                    .status("PENDING")
                    .build();
            MpesaTransaction savedTransaction = mpesaTransactionRepository.save(transaction);
            
            // Simulate callback after 4 seconds
            simulateCallbackAsync(checkoutRequestId, requestDto.getAmount(), 4000);
            
            return savedTransaction;
        }

        // Real Daraja Integration
        String token = getAccessToken();
        if (token == null) {
            throw new RuntimeException("Failed to authenticate with Safaricom Daraja");
        }

        String baseUrl = "production".equalsIgnoreCase(env) ? "https://api.safaricom.co.ke" : "https://sandbox.safaricom.co.ke";
        String url = baseUrl + "/mpesa/stkpush/v1/processrequest";

        String timestamp = new java.text.SimpleDateFormat("yyyyMMddHHmmss").format(new java.util.Date());
        String password = getPassword(timestamp);

        String formattedPhone = requestDto.getPhoneNumber().trim();
        if (formattedPhone.startsWith("0")) {
            formattedPhone = "254" + formattedPhone.substring(1);
        } else if (formattedPhone.startsWith("+")) {
            formattedPhone = formattedPhone.substring(1);
        }

        Map<String, Object> body = new java.util.HashMap<>();
        body.put("BusinessShortCode", Integer.parseInt(shortcode));
        body.put("Password", password);
        body.put("Timestamp", timestamp);
        body.put("TransactionType", "CustomerPayBillOnline");
        body.put("Amount", requestDto.getAmount().intValue());
        body.put("PartyA", Long.parseLong(formattedPhone));
        body.put("PartyB", Integer.parseInt(shortcode));
        body.put("PhoneNumber", Long.parseLong(formattedPhone));
        body.put("CallBackURL", callbackUrl);
        body.put("AccountReference", "ChamaSmart");
        body.put("TransactionDesc", "Chama Contribution");

        HttpHeaders headers = new HttpHeaders();
        headers.set("Authorization", "Bearer " + token);
        headers.setContentType(MediaType.APPLICATION_JSON);
        HttpEntity<Map<String, Object>> entity = new HttpEntity<>(body, headers);

        try {
            ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.POST, entity, Map.class);
            Map<String, Object> responseBody = response.getBody();
            
            if (response.getStatusCode().is2xxSuccessful() && responseBody != null) {
                String responseCode = (String) responseBody.get("ResponseCode");
                if ("0".equals(responseCode)) {
                    String checkoutRequestId = (String) responseBody.get("CheckoutRequestID");
                    String merchantRequestId = (String) responseBody.get("MerchantRequestID");
                    
                    MpesaTransaction transaction = MpesaTransaction.builder()
                            .checkoutRequestId(checkoutRequestId)
                            .merchantRequestId(merchantRequestId)
                            .user(user)
                            .chama(chama)
                            .amount(requestDto.getAmount())
                            .phoneNumber(formattedPhone)
                            .status("PENDING")
                            .build();
                    MpesaTransaction savedTransaction = mpesaTransactionRepository.save(transaction);
                    log.info("Daraja STK Push initiated successfully. CheckoutRequestID: {}", checkoutRequestId);
                    
                    // If local testing, Daraja won't be able to hit the callback URL, so we simulate it.
                    if (callbackUrl.contains("localhost") || env.equalsIgnoreCase("local")) {
                        simulateCallbackAsync(checkoutRequestId, requestDto.getAmount(), 15000);
                    }
                    
                    return savedTransaction;
                } else {
                    throw new RuntimeException("Daraja error: " + responseBody.get("ResponseDescription"));
                }
            } else {
                throw new RuntimeException("Failed to initiate STK Push: Invalid response from Daraja");
            }
        } catch (Exception e) {
            log.error("STK Push error: {}", e.getMessage());
            throw new RuntimeException("Failed to initiate M-Pesa STK Push: " + e.getMessage());
        }
    }

    private void simulateCallbackAsync(String checkoutRequestId, BigDecimal amount, long delayMs) {
        CompletableFuture.runAsync(() -> {
            try {
                Thread.sleep(delayMs);
                MpesaCallbackDto.Item amountItem = new MpesaCallbackDto.Item();
                amountItem.setName("Amount");
                amountItem.setValue(amount.intValue());

                MpesaCallbackDto.Item receiptItem = new MpesaCallbackDto.Item();
                receiptItem.setName("MpesaReceiptNumber");
                receiptItem.setValue("R" + UUID.randomUUID().toString().substring(0, 8).toUpperCase());

                MpesaCallbackDto.CallbackMetadata metadata = new MpesaCallbackDto.CallbackMetadata();
                metadata.setItem(java.util.Arrays.asList(amountItem, receiptItem));

                MpesaCallbackDto.StkCallback stkCallback = new MpesaCallbackDto.StkCallback();
                stkCallback.setCheckoutRequestID(checkoutRequestId);
                stkCallback.setResultCode(0);
                stkCallback.setResultDesc("Success - Simulated");
                stkCallback.setCallbackMetadata(metadata);

                MpesaCallbackDto.Body body = new MpesaCallbackDto.Body();
                body.setStkCallback(stkCallback);

                MpesaCallbackDto callbackDto = new MpesaCallbackDto();
                callbackDto.setBody(body);

                log.info("Executing simulated callback for CheckoutRequestID: {}", checkoutRequestId);
                applicationContext.getBean(MpesaService.class).processCallback(callbackDto);
            } catch (Exception e) {
                log.error("Error in simulated callback: {}", e.getMessage());
            }
        });
    }
    @Transactional
    public void processCallback(MpesaCallbackDto callbackDto) {
        if (callbackDto.getBody() == null || callbackDto.getBody().getStkCallback() == null) {
            log.error("Invalid M-Pesa callback payload received");
            return;
        }
        MpesaCallbackDto.StkCallback stkCallback = callbackDto.getBody().getStkCallback();
        String checkoutRequestId = stkCallback.getCheckoutRequestID();
        Integer resultCode = stkCallback.getResultCode();
        String resultDesc = stkCallback.getResultDesc();
        log.info("Processing M-Pesa Callback for CheckoutRequestID: {}, ResultCode: {}", checkoutRequestId, resultCode);
        MpesaTransaction transaction = mpesaTransactionRepository.findByCheckoutRequestId(checkoutRequestId)
                .orElseThrow(() -> new RuntimeException("MpesaTransaction not found for CheckoutRequestID: " + checkoutRequestId));
        transaction.setResultCode(resultCode);
        transaction.setResultDesc(resultDesc);
        if (resultCode == 0) {
            // Transaction Successful
            transaction.setStatus("COMPLETED");
            // Extract MpesaReceiptNumber from metadata
            String mpesaReceipt = "MPESA" + System.currentTimeMillis();
            if (stkCallback.getCallbackMetadata() != null && stkCallback.getCallbackMetadata().getItem() != null) {
                for (MpesaCallbackDto.Item item : stkCallback.getCallbackMetadata().getItem()) {
                    if ("MpesaReceiptNumber".equalsIgnoreCase(item.getName()) && item.getValue() != null) {
                        mpesaReceipt = item.getValue().toString();
                        break;
                    }
                }
            }
            transaction.setMpesaReceipt(mpesaReceipt);
            // Create Contribution record
            Contribution contribution = Contribution.builder()
                    .chama(transaction.getChama())
                    .user(transaction.getUser())
                    .amount(transaction.getAmount())
                    .reference(mpesaReceipt)
                    .status("COMPLETED")
                    .contributionType("REGULAR")
                    .isDeleted(false)
                    .build();
            Contribution savedContribution = contributionRepository.save(contribution);
            transaction.setContribution(savedContribution);
            // Update ChamaMember total contributions
            chamaMemberRepository.findByChamaChamaIdAndUserUserId(transaction.getChama().getChamaId(), transaction.getUser().getUserId())
                    .ifPresent(member -> {
                        member.setTotalContributions(member.getTotalContributions().add(transaction.getAmount()));
                        member.setLastContributionDate(ZonedDateTime.now());
                        chamaMemberRepository.save(member);
                    });
            // Update Chama current fund
            Chama chama = transaction.getChama();
            chama.setCurrentFund(chama.getCurrentFund().add(transaction.getAmount()));
            chamaRepository.save(chama);
            // Write immutable Financial Audit Log
            FinancialAuditLog auditLog = FinancialAuditLog.builder()
                    .user(transaction.getUser())
                    .transactionType("MPESA_CONTRIBUTION")
                    .amount(transaction.getAmount())
                    .chama(transaction.getChama())
                    .referenceId(savedContribution.getContributionId())
                    .description("Successful M-Pesa STK Push contribution. Receipt: " + mpesaReceipt)
                    .ipAddress("127.0.0.1") // Callback IP
                    .userAgent("Safaricom-Daraja-Webhook")
                    .build();
            auditLogRepository.save(auditLog);
            log.info("Successfully processed M-Pesa contribution for Chama ID: {}, Amount: {}", chama.getChamaId(), transaction.getAmount());
        } else {
            // Transaction Failed / Cancelled by user
            transaction.setStatus("FAILED");
            log.warn("M-Pesa transaction failed. ResultDesc: {}", resultDesc);
        }
        mpesaTransactionRepository.save(transaction);
    }
    @Transactional(readOnly = true)
    public List<ContributionSummaryDto> getContributionsByChamaId(Long chamaId) {
        log.info("Fetching contributions for chama ID: {}", chamaId);
        return contributionRepository.findByChamaChamaIdAndIsDeletedFalse(chamaId).stream()
                .map(ContributionSummaryDto::fromEntity)
                .collect(Collectors.toList());
    }
    @Transactional(readOnly = true)
    public List<ContributionSummaryDto> getMyContributions(Long userId) {
        log.info("Fetching contributions for user ID: {}", userId);
        return contributionRepository.findByUserUserIdAndIsDeletedFalse(userId).stream()
                .map(ContributionSummaryDto::fromEntity)
                .collect(Collectors.toList());
    }
}

